limitador-server/src/metrics.rs

```diff
@@ -0,0 +1,291 @@
+    }
+}
+
+#[derive(Debug, Clone)]
+struct SpanState {
+    group_times: HashMap<String, Timings>,
+}
+
+impl SpanState {
+    fn new(group: String) -> Self {
+        Self {
+            group_times: HashMap::from([(group, Timings::new())]),
+        }
+    }
+
+    fn increment(&mut self, group: &String, timings: Timings) -> &mut Self {
+        self.group_times
+            .entry(group.to_string())
+            .and_modify(|x| *x += timings)
+            .or_insert(timings);
+        self
+    }
+}
+
+pub struct MetricsGroup<F: Fn(Timings)> {
+    consumer: F,
+    records: Vec<String>,
+}
+
+impl<F: Fn(Timings)> MetricsGroup<F> {
+    pub fn new(consumer: F, records: Vec<String>) -> Self {
+        Self { consumer, records }
+    }
+}
+
+pub struct MetricsLayer<F: Fn(Timings)> {
+    groups: HashMap<String, MetricsGroup<F>>,
+}
+
+impl<F: Fn(Timings)> MetricsLayer<F> {
+    pub fn new() -> Self {
+        Self {
+            groups: HashMap::new(),
+        }
+    }
+
+    pub fn gather(mut self, aggregate: &str, consumer: F, records: Vec<&str>) -> Self {
+        // TODO(adam-cattermole): does not handle case where aggregate already exists
+        let rec = records.iter().map(|r| r.to_string()).collect();
+        self.groups
+            .entry(aggregate.to_string())
+            .or_insert(MetricsGroup::new(consumer, rec));
+        self
+    }
+}
+
+impl<S, F: Fn(Timings) + 'static> Layer<S> for MetricsLayer<F>
+where
+    S: Subscriber,
+    S: for<'lookup> LookupSpan<'lookup>,
+{
+    fn on_new_span(&self, _attrs: &Attributes<'_>, id: &Id, ctx: Context<'_, S>) {
+        let span = ctx.span(id).expect("Span not found, this is a bug");
+        let mut extensions = span.extensions_mut();
+        let name = span.name();
+
+        // if there's a parent
+        if let Some(parent) = span.parent() {
+            // if the parent has SpanState propagate to this span
+            if let Some(parent_state) = parent.extensions_mut().get_mut::<SpanState>() {
+                extensions.insert(parent_state.clone());
+            }
+        }
+
+        // if we are an aggregator
+        if self.groups.contains_key(name) {
+            if let Some(span_state) = extensions.get_mut::<SpanState>() {
+                // if the SpanState has come from parent and we must append
+                // (we are a second level aggregator)
+                span_state
+                    .group_times
+                    .entry(name.to_string())
+                    .or_insert(Timings::new());
+            } else {
+                // otherwise create a new SpanState with ourselves
+                extensions.insert(SpanState::new(name.to_string()))
+            }
+        }
+
+        if let Some(span_state) = extensions.get_mut::<SpanState>() {
+            // either we are an aggregator or nested within one
+            for group in span_state.group_times.keys() {
+                for record in &self.groups.get(group).unwrap().records {
```

Comment:
Could you use `expect` instead of `unwrap` here... just so avoid adding yet more of them to the code base until we address #280